src/ripple/overlay/impl/PeerImp.cpp

```diff
@@ -2851,4 +2861,36 @@ PeerImp::isHighLatency() const
     return latency_ >= Tuning::peerHighLatency;
 }
 
+void
+PeerImp::Metrics::add_message(std::uint64_t bytes)
+{
+    using namespace std::chrono_literals;
+
+    totalBytes += bytes;
+    accumBytes += bytes;
+    auto const actualInterval = clock_type::now() - intervalStart;
+
+    if (actualInterval >= 1s)
+    {
+        auto const avgBytes = accumBytes / std::chrono::duration_cast<std::chrono::seconds>(actualInterval).count();
+        rollingAvg[currIndex] = avgBytes;
+        rollingAvgBytes = std::accumulate(rollingAvg.begin(), rollingAvg.end(), 0ull) / rollingAvg.size();
+
+        currIndex = (currIndex + 1) % rollingAvg.size();
+        intervalStart = clock_type::now();
+        accumBytes = 0;
+    }
+}
+
+std::uint64_t
+PeerImp::Metrics::average_bytes() const {
+    return rollingAvgBytes;
```

Comment:
I know that 32-bit values are atomic, but this (and the next) probably needs to be protected by a lock.

Reply:
FWIW, please never assume that any word size (even an octet) is atomic.  In C++11 and greater _any_ race condition is undefined behavior.

Reply:
Wasn't assuming it was atomic, it's just not a critical code-path and didn't want to over complicate it. The only "bad thing" is you get a slightly older answer. But I can put it behind a lock for correctness :)

Reply:
>FWIW, please never assume that any word size (even an octet) is atomic. In C++11 and greater any race condition is undefined behavior.

I was talking about the underlying x86 semantics. But your point is well-made.

Reply:
I added a `shared_mutex` but still debating if we should just make the `uint64_t`'s atomic,